File: Ruby_version/Napakalaki/lib/numeric_bad_consequence.rb
# encoding: utf-8
# begin numeric_bad_consequence.rb

require_relative 'bad_consequence'

module NapakalakiGame
  class NumericBadConsequence < BadConsequence
    
    def initialize(t, l, nVisible, nHidden)
      super(t, l)
      @nVisibleTreasures = nVisible
      @nHiddenTreasures = nHidden
    end
    
    def getNVisibleTreasures
      @nVisibleTreasures  
    end

    def getNHiddenTreasures
      @nHiddenTreasures 
    end
    
    def isEmpty
      @nVisibleTreasures == 0 and @nHiddenTreasures == 0
    end
    
    def substractVisibleTreasure(t)
      if @nVisibleTreasures > 0
        @nVisibleTreasures -= 1 
      end
    end
    
    def substractHiddenTreasure(t)
      if @nHiddenTreasures > 0
        @nHiddenTreasures -= 1
      end
    end
    
    def adjustToFitTreasureLists(v, h)
      nV = v.size < @nVisibleTreasures ? v.size : @nVisibleTreasures
      nH = h.size < @nHiddenTreasures ? h.size : @nHiddenTreasures
      bC = NumericBadConsequence.new(@text, @levels, nV, nH)
    end
    
    def to_s
      super + "\n Tesoros visibles perdidos: #{@nVisibleTreasures}\n Tesoros ocultos perdidos: #{@nHiddenTreasures}"
    end
    
    def malRolloString
      super + "\n Tesoros visibles: #{@nVisibleTreasures}\n Tesoros ocultos: #{@nHiddenTreasures}"
    end
  end
end

# end numeric_bad_consequence.rb

File: Java_version/Napakalaki/src/NapakalakiGame/NumericBadConsequence.java
// begin NumericBadConsequence.java
package NapakalakiGame;

// Usar arrays
import java.util.ArrayList;

public class NumericBadConsequence extends BadConsequence {
    // Nº máximo de tesoros
    final static int MAXTREASURES = 10;
    // Nº de tesoros visibles perdidos
    private int nVisibleTreasures;
    // Nº de tesoros ocultos perdidos
    private int nHiddenTreasures;
    
    // Constructor tipo 1: tesoros no específcos
    public NumericBadConsequence(String t, int l, int nVisible, int nHidden) {
        super(t, l);
        this.nVisibleTreasures = nVisible;
        this.nHiddenTreasures = nHidden;
    }
    
    public int getNVisibleTreasures() {
        return nVisibleTreasures;
    }
    
    public int getNHiddenTreasures() {
        return nHiddenTreasures;
    }
    
    @Override
    public boolean isEmpty() {
        return nVisibleTreasures == 0 && nHiddenTreasures == 0;
    }
    
    @Override
    public void substractVisibleTreasure(Treasure t) {
        if (nVisibleTreasures > 0)
            nVisibleTreasures -= 1;            
    }
    
    @Override
    public void substractHiddenTreasure(Treasure t) {
        if (nHiddenTreasures > 0)
            nHiddenTreasures -= 1;  
    }
    
    @Override
    public BadConsequence adjustToFitTreasureLists(ArrayList<Treasure> v,
            ArrayList<Treasure> h) {
        int nV = v.size() < nVisibleTreasures ? v.size() : nVisibleTreasures;
        int nH = h.size() < nHiddenTreasures ? h.size() : nHiddenTreasures;
        BadConsequence bd = new NumericBadConsequence(getText(), getLevels(), 
                nV, nH);
        return bd;
    }
    
    @Override
    public String toString() {
        return super.toString() + "\n Tesoros visibles perdidos: "+
                Integer.toString(nVisibleTreasures) + 
                "\n Tesoros ocultos perdidos: " + 
                Integer.toString(nHiddenTreasures);
    }
    
    @Override
    public String toDiscard() {
        return "Visibles: " + Integer.toString(nVisibleTreasures) +
            "\n Ocultos: " + Integer.toString(nHiddenTreasures);
    }
}

// end NumericBadConsequence.java